---
title: "QC and filtering for conservative counts"
output:
  html_document: default
  html_notebook: default
---

QC and filtering for conservative counts.

```{r loading files, include=FALSE}

# Load necessary libraries

library(readr) # fast reading of large files
library(edgeR)
library(biomaRt)  # for gene information

library(knitr)   # for kable tables

# functions:
order_by_stages=function(counts,stage.=stage){
  nc=counts[ , grepl( "iPSC" , names( counts) ) ]  #takes columns whose name matches x
  
  #Do the same for the other stages
  
  for (s in stage.[-1])  {         
    
    i= counts[ , grepl( s , names( counts ) ) ]
    nc=cbind(nc,i)
  }
  
  return(nc)
}


trim_by_stage=function(dge){
  data=dge$counts
  data[data>0]=1 # convert to table where 0=0 and 1 equals any value above 0

  c=c() # initiate vector of numbers, to save rows to take out of matrix

  for (r in 1:nrow(data)){
   for (i in seq(1,24,3)){   # loop through stages 3 by 3 (to jump over samples of same stage)
    
     if (sum(data[r,c(i,i+1,i+2)]) > 2) {
      
        break
      }
      if (i==22){
      
        c=c(c,r)  # if it gets to the end and there are no valid values, append to vector
      }
    
    
   }
  
  }

rm(data)
return(c)

}


cc=read_tsv("/Users/Marta/Documents/WTCHG/DPhil/Data/Diff_v2/31.01.2017.Differentiation_v2.gene.counts.tsv")  # master object, is a list of matrices (24 elements, 1 per sample)
# each element of the list is a matrix with 57820 rows (gene ids) and 26 columns (gene id, length, counts, TPM, etc)


donors= c("Ad2.1","Ad3.1","Neo1.1")  # data comes from three donors

stage= c("iPSC", "DE", "PGT", "PFG", "PE", "EP","ENstage6", "ENstage7") # 8 stages
 

```
After loading the counts, I create a "counts" object, with genes as rows and samples as columns. 


```{r count object, echo=FALSE}

#conservative counts


counts_cc=as.data.frame(cc)   # make counts data frame
rm(cc)
rownames(counts_cc)=counts_cc[,1]
counts_cc=counts_cc[,c(-1:-2)]

rownames(counts_cc) <- gsub("(ENSG[0-9]+).*", "\\1", rownames(counts_cc)) # cut everything after the dot


```

Conservative counts initially has `r nrow(counts_cc)` genes, and looks like this: 

```{r conservative counts, echo=FALSE, tidy=TRUE}

counts_cc=order_by_stages(counts_cc)
colnames(counts_cc)=paste( rep(stage,each=3),rep(donors,8),sep="_" )

kable(counts_cc[1:2,1:6])

```

Now I will keep in my count object just the genes that still exist in ENSEMBL.

```{r ENSEMBL trimming, echo=FALSE}

#conservative counts
ensembl38 <- useMart("ENSEMBL_MART_ENSEMBL",dataset="hsapiens_gene_ensembl",host="www.ensembl.org")

all_ensembl_info_ensemblFound<- getBM(attributes=c('ensembl_gene_id', 'external_gene_name', 'gene_biotype', 'chromosome_name'),
                          filters = 'ensembl_gene_id', values = rownames(counts_cc), mart = ensembl38)
rownames(all_ensembl_info_ensemblFound) <- all_ensembl_info_ensemblFound$ensembl_gene_id


diff=setdiff(rownames(counts_cc),rownames(all_ensembl_info_ensemblFound)) # IDs lost (not found in GRCh38) , 5188 genes

ensembl37 = useMart(biomart="ENSEMBL_MART_ENSEMBL", host="grch37.ensembl.org", path="/biomart/martservice" ,dataset="hsapiens_gene_ensembl") 

all_ensembl_info_ensemblNotFound<- getBM(attributes=c('ensembl_gene_id', 'external_gene_name', 'gene_biotype', 'chromosome_name'),
                          filters = 'ensembl_gene_id', values = diff, mart = ensembl37)   # all info in GRCh37 of genes not found in GRCh38

# check by gene names if they exist in grch38


all_ensembl_info_ensemblNotFoundGRCh38<- getBM(attributes=c('ensembl_gene_id', 'external_gene_name', 'gene_biotype', 'chromosome_name'),
                          filters = 'external_gene_name', values = all_ensembl_info_ensemblNotFound$external_gene_name, mart = ensembl38) 
# there are duplicated ids. Check which ones are correct calling grch38 with ids

all_ensembl_info_ensemblNotFoundGRCh38Ids=getBM(attributes=c('ensembl_gene_id', 'external_gene_name', 'gene_biotype', 'chromosome_name'),
                          filters = 'ensembl_gene_id', values = all_ensembl_info_ensemblNotFoundGRCh38$ensembl_gene_id, mart =ensembl38) 
# remove those not in autosomes, as we have duplicate ids from weird chromosome assemblies

chr <- c(1:22)
all_ensembl_info_ensemblNotFoundGRCh38Ids <- all_ensembl_info_ensemblNotFoundGRCh38Ids[which(all_ensembl_info_ensemblNotFoundGRCh38Ids$chromosome_name %in% chr),]


keep1=all_ensembl_info_ensemblFound #keep genes in counts whose Ensembl geneID still exists in current genome build. 

# take genes not found in GRCh38 by id, that still exist in GRCh38 with different id
# genes that change id, to then save old and new id
toChangeGeneNames=intersect(all_ensembl_info_ensemblNotFoundGRCh38Ids$external_gene_name, all_ensembl_info_ensemblNotFound$external_gene_name)

old=all_ensembl_info_ensemblNotFound[which(  all_ensembl_info_ensemblNotFound$external_gene_name %in% toChangeGeneNames),c(1:2)]
colnames(old)=c("old_id","external_gene_name")
anyDuplicated(old)  # but there are duplicates! (weird genes with lots of gene ids)
old=old[which(!duplicated(old$external_gene_name)),]

new=all_ensembl_info_ensemblNotFoundGRCh38Ids[which(all_ensembl_info_ensemblNotFoundGRCh38Ids$external_gene_name %in% toChangeGeneNames),]
new=new[which(!duplicated(new$external_gene_name)),]


keep2=merge(old,new,by="external_gene_name") # merge old and new info
rownames(keep2)=keep2$old_id

# modify IDs in counts of those that change id
# toModify = which(rownames(counts_cc) %in%  keep2$old_id)  # select rows to change in counts
# keep2=keep2[rownames(counts_cc[toModify,]),]# same order as toModify
#rownames(counts_cc)[toModify] <- keep2[,"ensembl_gene_id"]  # several gene ids are duplicated, for some reason
# change just HNF1B
keep2=keep2[which(keep2$external_gene_name=="HNF1B"),]
rownames(counts_cc)[which(rownames(counts_cc)==keep2$old_id)] <- keep2$ensembl_gene_id # cange old id for new id
rownames(keep2)=keep2$ensembl_gene_id  # change rowname to new id

keep3=rbind(keep1,keep2[c(1,3:5)])  # join

dup=keep3[which(duplicated.data.frame(keep3)),] 
# x duplicated elements. Remove them

keep3=keep3[which(!duplicated.data.frame(keep3)),] # 52633


# merge info from found and not found that still exist with different iD in GRCh38
counts_cc=counts_cc[rownames(keep3),]  # 52633 genes remain

all_ensembl_info_cc = keep3

rm(keep1,keep2,keep3,all_ensembl_info_ensemblFound,all_ensembl_info_ensemblNotFound,all_ensembl_info_ensemblNotFoundGRCh38,all_ensembl_info_ensemblNotFoundGRCh38Ids)
```

After trimming,  conservative counts has `r nrow(counts_cc)` genes.

Now I want to keep in my count object genes with at least 1 cpm in at least one stage.


```{r cpm trimming, echo=FALSE}


group <- rep(stage,each=3)    #group by stages 

anyNA(counts_cc)  # remove NAs:

counts_cc=na.omit(counts_cc)

# remove from gene info genes not in counts


all_ensembl_info_cc = all_ensembl_info_cc[rownames(counts_cc),]  # 52632 genes remain

# create DGE object
dge_cc <- DGEList(counts=counts_cc,group=group,genes=all_ensembl_info_cc) 
all_ensembl_info_cc=all_ensembl_info_cc[rownames(dge_cc$counts),] # re-sort

# filter simply (remove genes with less that 3 columns with cpm<=1)
keep_cc <- rowSums(cpm(dge_cc)>1) >= 3   
dge_cc <- dge_cc[keep_cc, , keep.lib.sizes=FALSE]  # 17779 genes remain   

```
After trimming,  conservative counts has `r nrow(dge_cc)` genes.

Now I want to correct the problem from the previous QC: I want to remove genes in which at least one stage has counts for ALL 3 samples.

```{r more trimming, echo=FALSE}
# go through the table 3 by 3
# if there are more than 2 values in three columns above 0, go to next row.
# if not, go to next 3
# if gets to end of loop and not found the values, get that row. After loop, subset matrix excluding those rows


c=trim_by_stage(dge_cc)
fail_cc=100*(length(c)/nrow(dge_cc$counts))
genes_fail_cc=length(c)
ditch=dge_cc$counts[c,] # table to throw away

keep=!(rownames(dge_cc$counts) %in% rownames(ditch))

dge_cc <- dge_cc[keep, , keep.lib.sizes=FALSE] # counts now has 17776 genes
    
rm(c,keep,ditch)

```
After trimming, conservative counts has `r nrow(dge_cc)` genes. Conservative counts has lost  `r genes_fail_cc` genes (`r fail_cc`% from previous total).

Now I will 1) keep only lincRNA and protein coding genes, and 2) exclude mitochondrial genes and those in sex chromosomes.

``` {r even more trimming, echo=FALSE}
#remove genes that are not annotated as lincRNA or protein_coding
before=nrow(dge_cc)
dge_cc <- dge_cc[which(dge_cc$genes$gene_biotype == "protein_coding" | dge_cc$genes$gene_biotype == "lincRNA"),]


# now has 15810 genes

gene_lost_step_1_cc=before-nrow(dge_cc)
before=nrow(dge_cc)

# use only autosomal genes (not sex genes or mitocondrial genes)
chr <- c(1:22)
dge_cc <- dge_cc[which(dge_cc$genes$chromosome_name %in% chr),]

# now has 15221 genes


```
Conservative counts has `r nrow(dge_cc)` genes, and lost `r gene_lost_step_1_cc` genes on the first trim, and `r gene_lost_step_2_cc` genes on the second. 


And the library sizes for conservative counts:

``` {r library cc, echo=FALSE}
kable(dge_cc$samples[,-3])
```

We can then save the dge object we want to keep (in our case, conservative counts), using the following code:


``` {r saving cc, echo=FALSE}
setwd("/Users/Marta/Documents/WTCHG/DPhil/Data/Diff_v2/session_objects")

    save(dge_cc, file = "dge_cc.xz" , compress="xz")
```


